File: app/src/features/flight-planner/utils/droneSpecs.ts
import type { DroneSpec, SelectOption } from "@/features/flight-planner/types";

const DRONE_SPECS: Record<string, DroneSpec> = {
  "Mini 4 Pro": {
    model: "Mini 4 Pro",
    sensorWidthMm: 9.6,
    sensorHeightMm: 7.2,
    focalLengthMm: 6.72,
    imageWidthPx: 8064,
    imageHeightPx: 6048,
    maxSpeedMs: 16,
    batteryTimeMin: 34,
  },
  "Mini 5 Pro": {
    model: "Mini 5 Pro",
    image: "/mini-5-pro.png",
    sensorWidthMm: 13.2,
    sensorHeightMm: 8.8,
    focalLengthMm: 7.33,
    imageWidthPx: 8192,
    imageHeightPx: 6144,
    maxSpeedMs: 18,
    batteryTimeMin: 36,
  },
  "Air 3": {
    model: "Air 3",
    sensorWidthMm: 9.6,
    sensorHeightMm: 7.2,
    focalLengthMm: 6.7,
    imageWidthPx: 8064,
    imageHeightPx: 6048,
    maxSpeedMs: 21,
    batteryTimeMin: 42,
  },
  "Mavic 3": {
    model: "Mavic 3",
    sensorWidthMm: 17.3,
    sensorHeightMm: 13,
    focalLengthMm: 12.3,
    imageWidthPx: 5280,
    imageHeightPx: 3956,
    maxSpeedMs: 21,
    batteryTimeMin: 46,
  },
  "Phantom 4": {
    model: "Phantom 4",
    sensorWidthMm: 13.2,
    sensorHeightMm: 8.8,
    focalLengthMm: 8.8,
    imageWidthPx: 5472,
    imageHeightPx: 3648,
    maxSpeedMs: 20,
    batteryTimeMin: 28,
  },
};

export function getDroneOptions(): SelectOption[] {
  return Object.keys(DRONE_SPECS).map((model) => ({
    label: model,
    value: model,
  }));
}

/** Especificações locais quando o catálogo API não está disponível ou não contém o nome. */
export function getDroneSpec(model: string): DroneSpec {
  const direct = DRONE_SPECS[model]
  if (direct) return direct
  const k = Object.keys(DRONE_SPECS).find((x) => x.toLowerCase() === model.trim().toLowerCase())
  if (k) return DRONE_SPECS[k]!
  return DRONE_SPECS['Mini 4 Pro']!
}

File: backend/app/core/flight/drone_specs.py
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class DroneSpec:
    model: str
    sensor_width_mm: float
    sensor_height_mm: float
    focal_length_mm: float
    image_width_px: int
    image_height_px: int


DRONE_SPECS: dict[str, DroneSpec] = {
    "mini-4-pro": DroneSpec(
        model="Mini 4 Pro",
        sensor_width_mm=9.6,
        sensor_height_mm=7.2,
        focal_length_mm=6.7,
        image_width_px=8064,
        image_height_px=6048,
    ),
    "mini-5-pro": DroneSpec(
        model="Mini 5 Pro",
        sensor_width_mm=13.2,
        sensor_height_mm=8.8,
        focal_length_mm=7.33,
        image_width_px=8192,
        image_height_px=6144,
    ),
    "air-3": DroneSpec(
        model="Air 3",
        sensor_width_mm=9.6,
        sensor_height_mm=7.2,
        focal_length_mm=6.7,
        image_width_px=8064,
        image_height_px=6048,
    ),
    "mavic-3": DroneSpec(
        model="Mavic 3",
        sensor_width_mm=17.3,
        sensor_height_mm=13.0,
        focal_length_mm=12.3,
        image_width_px=5280,
        image_height_px=3956,
    ),
    "phantom-4": DroneSpec(
        model="Phantom 4",
        sensor_width_mm=13.2,
        sensor_height_mm=8.8,
        focal_length_mm=8.8,
        image_width_px=5472,
        image_height_px=3648,
    ),
}


def get_specs(model: str) -> DroneSpec:
    key = model.strip().lower()
    if key not in DRONE_SPECS:
        raise ValueError(f"Unsupported drone model: {model}")
    return DRONE_SPECS[key]
